### PR TITLE
Use ErrorList for experiment validator

### PR DESCRIPTION
**Pull request by ckcd**
**What this PR does / why we need it**:
For experiment validator, return errors after all fields are verified.

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubeflow/katib/issues/2302

**Checklist:**

- [x] [Docs](https://www.kubeflow.org/docs/components/katib/) included if any changes are user facing

### CODE REVIEW

**Comment by tenzen-y** [comment]
> /retest

This command should not work well.

**Comment by ckcd** [comment]
> This command should not work well.

@tenzen-y Thanks... seems no way for me to retry. I just push a new commit to re-trigger the tests.

Also please help to review this, thanks.
/assign @tenzen-y

**Comment by ckcd** [comment]
Also not sure why some test will fail randomly. They all runs well in my local. cc @tenzen-y

**Comment by tenzen-y** [comment]
> Details

That is a flaky test: https://github.com/kubeflow/katib/issues/2269

File: pkg/webhook/v1beta1/experiment/validator/validator.go [hunk 1]
```diff
@@ -108,194 +126,217 @@ func (g *DefaultValidator) ValidateExperiment(instance, oldInst *experimentsv1be
...
 	// Check if trialParameters exists
 	if trialTemplate.TrialParameters == nil {
-		return fmt.Errorf("spec.trialTemplate.trialParameters must be specified")
+		return append(allErrs, field.Required(trialTemplatePath.Child("trialParameters"), "must be specified"))
 	}
 
 	// Check if trialSpec or configMap exists
 	if trialTemplate.TrialSource.TrialSpec == nil && trialTemplate.TrialSource.ConfigMap == nil {
-		return fmt.Errorf("spec.trialTemplate.trialSpec or spec.trialTemplate.configMap must be specified")
+		return append(allErrs, field.Required(trialTemplatePath.Child("TrialSource"), "spec.trialTemplate.trialSpec or spec.trialTemplate.configMap must be specified"))
```
Comment (andreyvelich) [inline pkg/webhook/v1beta1/experiment/validator/validator.go:319]: Same here

**Review by andreyvelich — COMMENTED** [review]
Thank you for doing this @ckcd!
I left a few comments.

**Comment by ckcd** [comment]
@andreyvelich Thanks for your reply! I have made the changes according to your opinion. PTAL again, thanks.

File: pkg/webhook/v1beta1/experiment/validator/validator.go [hunk 2]
```diff
@@ -420,79 +472,93 @@ func validatePatchJob(runSpec *unstructured.Unstructured, job interface{}, jobTy
...
-			return fmt.Errorf("file path where metrics file exists is required by .spec.metricsCollectorSpec.source.fileSystemPath.path")
+			allErrs = append(allErrs, field.Required(metricsSourcePath.Child("fileSystemPath").Child("path"),
+				"file path where metrics file exists is required by .spec.metricsCollectorSpec.source.fileSystemPath.path"))
 		}
 		// Format
 		fileFormat := mcSpec.Source.FileSystemPath.Format
 		if fileFormat != commonapiv1beta1.TextFormat && fileFormat != commonapiv1beta1.JsonFormat {
-			return fmt.Errorf("format of metrics file is required by .spec.metricsCollectorSpec.source.fileSystemPath.format")
+			allErrs = append(allErrs, field.Required(metricsSourcePath.Child("fileSystemPath").Child("format"),
+				"format of metrics file is required by .spec.metricsCollectorSpec.source.fileSystemPath.format"))
```
Comment (andreyvelich) [inline pkg/webhook/v1beta1/experiment/validator/validator.go:503]: @ckcd How the error message will look like that we send to the users ?
Do we need to simplify it as follows:
`format of metrics file is required for metrics collector` ?

Comment (ckcd) [reply pkg/webhook/v1beta1/experiment/validator/validator.go]: @andreyvelich Yes the error message looks a little redundant:
```
[spec.metricsCollectorSpec.source.fileSystemPath.format: Required value: format of metrics file is required by .spec.metricsCollectorSpec.source.fileSystemPath.format]
```

I have made changes to remove redundant information according to your suggestion. Thanks.

File: pkg/webhook/v1beta1/experiment/validator/validator.go [hunk 3]
```diff
@@ -420,79 +472,93 @@ func validatePatchJob(runSpec *unstructured.Unstructured, job interface{}, jobTy
...
 		for _, mFormat := range mcSpec.Source.Filter.MetricsFormat {
 			if _, err := regexp.Compile(mFormat); err != nil {
-				return fmt.Errorf("invalid %q in .spec.metricsCollectorSpec.source.filter: %v", mFormat, err)
+				allErrs = append(allErrs, field.Invalid(metricsSourcePath.Child("filter").Child("metricsFormat"),
+					mFormat, fmt.Sprintf("invalid %q in .spec.metricsCollectorSpec.source.filter: %v", mFormat, err)))
 			} else {
 				if !mustTwoBracket.MatchString(mFormat) {
-					return fmt.Errorf("invalid %q in .spec.metricsCollectorSpec.source.filter: two top subexpressions are required", mFormat)
+					allErrs = append(allErrs, field.Invalid(metricsSourcePath.Child("filter").Child("metricsFormat"),
+						mFormat, fmt.Sprintf("invalid %q in .spec.metricsCollectorSpec.source.filter: two top subexpressions are required", mFormat)))
```
Comment (andreyvelich) [inline pkg/webhook/v1beta1/experiment/validator/validator.go:555]: Do you also want to refactor error messages for `metricsCollectorSpec` given that we we pass the Child info in the error ?

Comment (andreyvelich) [reply pkg/webhook/v1beta1/experiment/validator/validator.go:555]: E.g. the error message might look as follows:
```
[spec.metricsCollectorSpec.source.fileSystemPath.format: Required value: format of metrics file is required]
```

**Comment by andreyvelich** [comment]
Hi @ckcd, do you have some time to work on this PR so we can merge it ?

File: pkg/webhook/v1beta1/experiment/validator/validator.go (hunk 3 above)
Comment (ckcd) [reply pkg/webhook/v1beta1/experiment/validator/validator.go]: @andreyvelich done, sorry for my delay

**Comment by ckcd** [comment]
@andreyvelich Sorry for my delay, I have modified all the`metricsCollectorSpec` related error message according to your suggestion. PTAL, and please let me know if you have any comments.
Thanks.

**Comment by andreyvelich** [comment]
Please rebase your PR @ckcd, it should fix the E2Es

**Comment by ckcd** [comment]
@andreyvelich sure, I already rebase it.

**Review by andreyvelich — COMMENTED** [review]
Thanks for doing this @ckcd!
/lgtm
/assign @tenzen-y @johnugeorge

**Comment by ckcd** [comment]
@tenzen-y @johnugeorge could you take a look at this? thanks.

**Review by andreyvelich — COMMENTED** [review]
I think, we should be ready to merge this.
Thank you for your contribution @ckcd 🎉 
/lgtm
/approve

**Pull request merged**